Add progressive enhancement in OOP post (#296)
epiverse-trace/epiverse-trace.github.io: posts/progressive-enhancement/index.qmd

---

> title: "Improving Ecosystem Interoperability Iteratively via Progressive Enhancement"

Reviewer: I'm just realising we are inconsistent with the use of capitalisation and sentence case for our titles. We should probably harmonise this in a separate issue/PR.

author:
  - name: "Hugo Gruson"
    orcid: "0000-0002-4094-1476"
date: "2024-07-04"
categories: [R, interoperability, S3, progressive enhancement, ecosystem, lifecycle, object orientation, DOI]
format:
  html: 
    toc: true
---

> We are continuing our post series on S3 and interoperability.

Reviewer: In order not to confuse it with other S3's in the programming landscape for example, the [S3 programming language](https://en.wikipedia.org/wiki/S3_(programming_language)#:~:text=S3%20is%20a%20structured%2C%20imperative,for%20its%202900%20Series%20mainframes)
```suggestion
We are continuing our post series on S3 object orientation and interoperability in R.
```

We have previously discussed [what makes a good S3 class and how to choose a good parent for it, as well as when to write or not write a custom method](../parent-class).
We have highlighted in particular how classes inheriting from data.frames can simplify user experience because of familiarity, and reduce developer workload due to the pre-existing S3 methods.

We have detailed how to improve compatibility with the tidyverse by explaining:

- [how functions taking data.frames or data.frames subclass should also allow compatibility with tibble, which can be done in a few steps](https://hugogruson.fr/posts/compa-tibble/)
- [how to ensure class attributes are preserved whenever possible while using dplyr functions](../extend-dataframes).

Here, we are going to explore how to start adding support in the ecosystem for the new S3 classes while minimizing user-facing breaking changes.
We have previously delved into this topic with our post ["Convert Your R Function to an S3 Generic: Benefits, Pitfalls & Design Considerations"](../s3-generic) and this is a wider and higher-level view of the same topic.

> The strategy presented here is the variation of a common concept in web development and the web ecosystem: [progressive enhancement](https://developer.mozilla.org/en-US/docs/Glossary/Progressive_Enhancement). This philosophy aims to support browsers with limited range of features, while allowing a slightly richer experience for browsers with extra features.

Reviewer: ```suggestion
The strategy presented here is the variation of a common concept in web development and the web ecosystem: [progressive enhancement](https://developer.mozilla.org/en-US/docs/Glossary/Progressive_Enhancement). This philosophy aims to support browsers with a common set of essential features, and even richer features for browser with the most recent updates.
```

It makes sense to think about this philosophy with the prism of introducing new classes to a new software ecosystem as it has the similar constraints of multiple stakeholders with different interests and timelines.

> The application of progressive enhancement in this context means that users or packages that did not (yet) adopt the new classes are not penalized compared to users or packages that did.

Reviewer: ```suggestion
The application of progressive enhancement in this context means that users or packages that have not (yet) adopted the new classes are not penalized compared to users or packages that have.
```

In this case, let's consider a function that was previously returning an unclassed `data.frame` with patient-level data:

```{r}
create_patient_dataset <- function(n_patients = 10) {

  data <- data.frame(
    patient_id = seq_len(n_patients),
    age = sample(18:99, n_patients, replace = TRUE)
  )

  return(data)

}
```

We want to start returning a `linelist` object. Because `linelist` objects are `data.frame`s (or `tibble`s) with an extra `attr`, it can be done in a transparent way:

```{r}
create_patient_dataset <- function(n_patients = 10) {

  data <- data.frame(
    patient_id = seq_len(n_patients),
    age = sample(18:99, n_patients, replace = TRUE)
  )

  data <- linelist::make_linelist(
    data,
    id = "patient_id",
    age = "age"
  )

  return(data)

}

inherits(data, "data.frame")
```

For a more realistic example, you can also see the work in progress to integrate the [new `contactmatrix` standard format](https://github.com/socialcontactdata/contactmatrix) for social contact data to the [contactdata package](https://github.com/Bisaloo/contactdata).

> This is however only true if code in downstream analysis follows good practices [^1]. If existing code was testing equality of the class to a certain value, it will break when the new class value is appended. This is described in a [post on the R developer blog, when base R was adding a new `array` class value to `matrix` objects](https://developer.r-project.org/Blog/public/2019/11/09/when-you-think-class.-think-again/index.html). Class inheritance should never be tested via `class(x) == "some_class"`. Instead, `inherits(x, "some_class")` or `is(x, "some_class")` should be used to future-proof the code and allow appending an additional in the future.

Reviewer: ```suggestion
This is however only true if code in downstream analysis follows good practices in checking for the class of an object [^1]. If existing code was testing equality of the class to a certain value, it will break when the new class value is appended. This is described in a [post on the R developer blog, when base R was adding a new `array` class value to `matrix` objects](https://developer.r-project.org/Blog/public/2019/11/09/when-you-think-class.-think-again/index.html). Class inheritance should never be tested via `class(x) == "some_class"`. Instead, `inherits(x, "some_class")` or `is(x, "some_class")` should be used to future-proof the code and allow appending an additional in the future.
```

[^1]: This is now [enforced in R packages by R CMD check](https://github.com/r-devel/r-svn/commit/77ebdff5adc200dfe9bc850bc4447088830d2ee0), and via the [`class_equals_linter()`](https://lintr.r-lib.org/reference/class_equals_linter.html)  in the [lintr package](https://lintr.r-lib.org/).

## Conclusion

Object oriented programming and S3 classes offer a convenient way to iteratively add interoperability in the ecosystem in a way that is minimally disruptive to users and developers.
Newly classed input support can be added via custom methods (after converting the existing function to a generic if necessary).

> Newly classed output support can be added via progressive enhancement, by ensuring that the new class is a subclass of the old one, that downstream code uses good practices.

Reviewer: The statement "that downstream code uses good practice" doesn't follow smoothly after the command but I'm unsure of what you're trying to say.

---
title: "Improving Ecosystem Interoperability Iteratively via Progressive Enhancement"
author:
  - name: "Hugo Gruson"
    orcid: "0000-0002-4094-1476"
date: "2024-07-04"
categories: [R, interoperability, S3, progressive enhancement, ecosystem, lifecycle, object orientation, DOI]
format:
  html: 
    toc: true
---

> We are continuing on post series on S3 and interoperability.

Reviewer: ```suggestion
We are continuing our post series on S3 and interoperability.
```

> We have previously discussed [what makes a good S3 class and how to choose a good parent for it, as well as when to write or not write a custom method](../parent-class).

Reviewer: ```suggestion
We have previously discussed [what makes a good S3 class and how to choose a good parent for it, as well as when to write a custom method](../parent-class).
```

> We have previously discussed [what makes a good S3 class and how to choose a good parent for it, as well as when to write or not write a custom method](../parent-class).

Reviewer: ```suggestion
We have previously discussed [what makes a good S3 class and how to choose a good parent for it, as well as when to write a custom method](../parent-class).
```

> We have highlighted in particular how classes inheriting from data.frames can simplify user experience, and reduce developer workload.

Reviewer: ```suggestion
We have highlighted in particular how classes inheriting from data.frames can simplify user experience because of familiarity, and reduce developer workload.
```

We have detailed how to improve compatibility with the tidyverse by explaining:

> - [how functions taking data.frames or data.frames subclass can easily and should also allow compatibility with tibble](https://hugogruson.fr/posts/compa-tibble/)

Reviewer: ```suggestion
- [how functions taking data.frames or data.frames subclass should also allow compatibility with tibble, which can easily be done](https://hugogruson.fr/posts/compa-tibble/)
```

> - [how to ensure class attribute is preserved whenever possible while using dplyr functions](../extend-dataframes).

Reviewer: ```suggestion
- [how to ensure class attributes are preserved whenever possible while using dplyr functions](../extend-dataframes).
```

> Here, we are going to explore how to actually start adding support in the ecosystem for the newly S3 classes while minimizing user-facing breaking changes.

Reviewer: ```suggestion
Here, we are going to explore how to actually start adding support in the ecosystem for the new S3 classes while minimizing user-facing breaking changes.
```

> We have previously delved into this topic with our previous post ["Convert Your R Function to an S3 Generic: Benefits, Pitfalls & Design Considerations"](../s3-generic) and this is a wider and higher-level view of the same topic.

Reviewer: ```suggestion
We have previously delved into this topic with our post ["Convert Your R Function to an S3 Generic: Benefits, Pitfalls & Design Considerations"](../s3-generic) and this is a wider and higher-level view of the same topic.
```

> The strategy presented here is the declination of a common concept in web development and the web ecosystem: [progressive enhancement](https://developer.mozilla.org/en-US/docs/Glossary/Progressive_Enhancement). This philosophy aims to support browsers with limited range of features, while allowing a slightly richer experience for browsers with extra features.

Reviewer: Did you mean to use "declination" here?

It makes sense to think about this philosophy with the prism of introducing new classes to a new software ecosystem as it has the similar constraints of multiple stakeholders with different interests and timelines.
The application of progressive enhancement in this context means that users or packages that did not (yet) adopt the new classes are not penalized compared to users or packages that did.

## Adding class support to function inputs via progressive enhancement

The goal here is to allow functions to accept the new classes as inputs, while keeping the old behaviour unchanged for unclassed objects (or with a different class than the new one).

This can conveniently be done in an almost transparent way by converting the old function to the S3 generic, and using the default method to handle the old behaviour. The practical steps, and minor caveats, have been previously described in the post ["Convert Your R Function to an S3 Generic: Benefits, Pitfalls & Design Considerations"](../s3-generic).

![A before / after type image showing the conversion of a function to a generic with a default method keeping the exisiting behaviour.](convert_to_generic.svg)

> If the function was already a generic, even steps are required: a new method for the new class should be added, leaving everything else unchanged.

Reviewer: ```suggestion
If the function was already a generic, then a new method for the new class should be added, leaving everything else unchanged.
```
